apps/hash-graph/lib/type-fetcher/src/fetcher_server.rs

```diff
@@ -0,0 +1,184 @@
+            let client = http_client.clone();
+            let next_url = state.queue.pop_front();
+
+            let Some(url) = next_url else { return Ok(None) };
+            let response = fetch_ontology_type(client, url).await?;
+
+            let uris = match response.clone() {
+                FetchedOntologyType::EntityType(schema) => {
+                    let entity_type: EntityType = schema.try_into().map_err(|error| {
+                        tracing::error!(error=?error, "Couldn't convert schema to Entity Type");
+                        FetcherError::TypeParsingError
+                    })?;
+                    traverse_entity_type_references(&entity_type)
+                }
+                FetchedOntologyType::PropertyType(schema) => {
+                    let property_type: PropertyType = schema.try_into().map_err(|error| {
+                        tracing::error!(error=?error, "Couldn't convert schema to Property Type");
+                        FetcherError::TypeParsingError
+                    })?;
+
+                    traverse_property_type_references(&property_type)
+                }
+                FetchedOntologyType::DataType(schema) => {
+                    let data_type: DataType = schema.try_into().map_err(|error| {
+                        tracing::error!(error=?error, "Couldn't convert schema to Data Type");
+                        FetcherError::TypeParsingError
+                    })?;
+
+                    traverse_data_type_references(&data_type)
+                }
+            };
+
+            for uri in uris {
+                // TODO: reconsider how we should handle these "builtin" external types.
+                if uri.starts_with("https://blockprotocol.org/@blockprotocol/types/") {
+                    continue;
+                }
+
+                if !state.seen.contains(&uri) {
+                    state.seen.insert(uri.clone());
+                    state.queue.push_back(uri);
+                }
+            }
+
+            Ok(Some((response, state)))
+        },
+    )
+    .try_collect::<Vec<_>>()
+    .await?;
+
+    Ok(TypeFetchResponse::new(res))
+}
+
+pub async fn fetch_ontology_type(
+    client: Client,
+    url: String,
+) -> Result<FetchedOntologyType, FetcherError> {
+    let resp = client
+        .get(&url)
+        .header(ACCEPT, "application/json")
+        .header(USER_AGENT, "HASH Graph")
+        .send()
+        .await
+        .map_err(|err| {
+            tracing::error!(error=?err, url=&url, "Could not fetch ontology type");
+            FetcherError::NetworkError
+        })?
+        .json::<FetchedOntologyType>()
+        .await
+        .map_err(|err| {
+            tracing::error!(error=?err, url=&url, "Could not deserialize response");
+            FetcherError::SerializationError(format!("Error deserializing {url}: {err:#?}"))
+        })?;
+
+    Ok(resp)
+}
+
+fn traverse_entity_type_references(entity_type: &EntityType) -> Vec<String> {
```

Comment:
I did try to do that, but since our reference methods on the ontology types return owned values (`HashSet` and `HashMap`s), we return references to temporary values in the iterator, do you have an idea for how we could get around this, or should we not bother?

Reply:
Done in c51a34b34b93aab1f0a8c1a05759aa261009bc02 but we couldn't quite go with only refs